Python/libraries/recognizers-date-time/recognizers_date_time/date_time/base_time_zone.py

```diff
@@ -0,0 +1,152 @@
+    @property
+    @abstractmethod
+    def direct_utc_regex(self) -> regex:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def location_time_suffix_regex(self) -> regex:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def location_matcher(self) -> StringMatcher:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def time_zone_matcher(self) -> StringMatcher:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def ambiguous_time_zone_list(self) -> List[str]:
+        raise NotImplementedError
+
+
+class BaseTimeZoneExtractor(DateTimeZoneExtractor):
+    @property
+    def extractor_type_name(self) -> str:
+        return Constants.SYS_DATETIME_TIME
+
+    def __init__(self, config: TimeZoneExtractorConfiguration):
+        self.config = config
+
+    def extract(self, source: str, reference: datetime = None) -> List[ExtractResult]:
+        from .utilities import merge_all_tokens
+        tokens: List[Token] = list()
+        normalized_text = QueryProcessor.remove_diacritics(source)
+        tokens.append(self.match_timezones(normalized_text))
+        tokens.append(self.match_location_times(normalized_text, tokens))
+        return merge_all_tokens(tokens, source, self.extractor_type_name)
+
+    def remove_ambiguous_time_zone(self, extract_result: List[ExtractResult]) -> List[ExtractResult]:
+        return [item for item in extract_result if self.config.ambiguous_time_zone_list in item.text]
+
+    def match_location_times(self, text: str, tokens: List[Token]) -> List[Token]:
+        result: List[Token] = list()
+
+        if not self.config.location_time_suffix_regex:
+            return result
+
+        time_match = list(regex.finditer(self.config.location_time_suffix_regex, text))
+
+        # Before calling a Find() in location matcher, check if all the matched suffixes by
+        # LocationTimeSuffixRegex are already inside tokens extracted by TimeZone matcher.
+        # If so, don't call the Find() as they have been extracted by TimeZone matcher, otherwise, call it.
+        is_all_suffix_inside_tokens = True
+        for match in time_match:
+            is_inside = False
+            for token in tokens:
+                if token.start <= match.index and token.end >= match.index + match.lenght:
+                    is_inside = True
+                    break
+                if not is_inside:
+                    is_all_suffix_inside_tokens = False
+
+                if not is_all_suffix_inside_tokens:
+                    break
+
+        if len(time_match) != 0 and not is_all_suffix_inside_tokens:
+            last_match_index = time_match[len(time_match)-1]
+            matches = regex.finditer(self.config.location_matcher, text[0: last_match_index])
+            location_matches = MatchingUtil.remove_sub_matches(regex.finditer(self.config.location_matcher, text[0: last_match_index]))
+
```

Comment:
Why you are not using 'matches'? Here you have to pass 'matches' to _remove_sub_matches()_

Reply:
Nice one. I did not saw it. Fixed.